### ISSUE TITLE
When building a database the names for some of the addresses are not cached in the tag cache

### ISSUE THREAD

**Issue by arizvisa** [arizvisa/ida-minsc#122]
As reported by a user, in some cases when building the tag cache from a fresh database, a new `__name__` attribute does not get its reference tracked in the tag cache. This was actually done by design so that IDA's default names that it assigns during analysis are ignored when querying a database. This way querying things with `database.select` will only select things that you've actually touched. That's why things like extra comments, names, and type information have their tags double-underscored (`__extra_prefix__` and `__extra_suffix__`, `__name__`, and `__typeinfo__`).

Despite this being intentional, this is not consistent with how people expect tags to work since getting the tag for an address will always return these dictionary keys with their values. The `custom.tagfix` module actually indexes __everything__, so as a result the history that I'm trying to preserve is explicitly lost when rebuilding the cache.

**Comment by arizvisa** [comment]
I'll need to figure out whether I should fix this so that it's always consistent, or I should keep the original logic which is apparently only useful to me. Originally the idea was that anything which is double-underscored is an implicit tag and is never indexed unless it's being explicitly set by the user. This is true for most things...except for the `__color__` tag (because I forgot about it and never use it), and the `__typeinfo__` tag (because I didn't think those hooks actually worked).

At the very least, this characteristic should be documented so that people who're actually using the tagging interface have knowledge about it. Some research should also be done to see if we can hook when the `__typeinfo__` for an address and `__color__` has changed so that we can update the cache when a user touches it.

**Comment by arizvisa** [comment]
I guess I'm going to create a separate branch for making it consistent, and then test whether it makes sense to retain the original functionality or to just index everything.

**Comment by arizvisa** [comment]
PR #125 is an implementation where the cache keeps track of all implicit tags during analysis and tagfix. Personally I don't like it...

**Comment by arizvisa** [comment]
Okay. So if we keep track of some of the implicit tags the user may modify, then we can use these hooks.

Used when changing the function prototype or the type for a data item:
```
# changing_ti(ea, new_type, new_fnames)
ui.hook.idb.add('changing_ti', Fchanging_ti, 1000)
# ti_changed(ea, type, fnames)
ui.hook.idb.add('ti_changed', Fchanged_ti, 1000)
```
Used when changing an instruction operand type:
```
# changing_op_type(ea, n, opinfo)
ui.hook.idb.add('changing_op_type', Fchanging_opt, 1000)
# op_type_changed(ea, n)
ui.hook.idb.add('op_type_changed', Fchanged_opt, 1000)
```
Used when changing colors:
```
# item_color_changed(ea, color=idaapi.COLOR_DEFAULT)
ui.hook.idb.add('item_color_changed', item_color_changed, 1000)
```

I haven't been able to figure out what "changing_op_ti" and "op_ti_changed" are used for. I was able to get it to work once, but I was testing things and thus wasn't ready to recall what hook was being executed.

**Comment by arizvisa** [comment]
For the record, this is what I'm using to test hooks:
```
def Flogger(variable, name, eaIndex):
    def F(*args, **kwargs):
        ea = args[eaIndex]
        print(hex(ea), 'hook', name)
        variable.setdefault(ea, []).append(name)
    return F
a={}

Fchanging_ti = Flogger(a, 'changing_ti', 0)
Fchanged_ti = Flogger(a, 'ti_changed', 0)
ui.hook.idb.add('changing_ti', Fchanging_ti, 1000)
ui.hook.idb.add('ti_changed', Fchanged_ti, 1000)
ui.hook.idb.disable('changed_ti'), ui.hook.idb.disable('ti_changed')

Fchanging_opti = Flogger(a, 'changing_op_ti', 0)
Fchanged_opti = Flogger(a, 'op_ti_changed', 0)
ui.hook.idb.add('changing_op_ti', Fchanging_opti, 1000)
ui.hook.idb.add('op_ti_changed', Fchanged_opti, 1000)

Fchanging_opt = Flogger(a, 'changing_op_type', 0)
Fchanged_opt = Flogger(a, 'op_type_changed', 0)
ui.hook.idb.add('changing_op_type', Fchanging_opt, 1000)
ui.hook.idb.add('op_type_changed', Fchanged_opt, 1000)
ui.hook.idb.disable('changing_op_type'), ui.hook.idb.disable('op_type_changed')

def item_color_changed(ea, color):
    if ea == idaapi.COLOR_DEFAULT:
        print(hex(ea), 'color', 'removed')
    else:
        print(hex(ea), 'color', hex(color))
    return

ui.hook.idb.add('item_color_changed', item_color_changed, 1000)
```

**Comment by arizvisa** [comment]
Just pushed a branch. Now if my connection will stop dropping packets so I could transfer at >5kb/s, then maybe I could create the fucking PR...

**Comment by arizvisa** [comment]
PR #126 adds support for caching the implicit tag types that the user has applied. I definitely prefer this one rather than the previous one I submitted.

**Comment by arizvisa** [comment]
This issue is fixed by PR #126 which was just merged. Closing this issue.
